gxtest/copyfilter.cpp

```diff
@@ -0,0 +1,562 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <fmt/format.h>
+
+#include <ogcsys.h>
+#include <wiiuse/wpad.h>
+#include "common/hwtests.h"
+#include "gxtest/cgx.h"
+#include "gxtest/cgx_defaults.h"
+#include "gxtest/util.h"
+
+// Use all copy filter values (0-63*3), instead of only 64
+#define FULL_COPY_FILTER_COEFS true
+// Use all gamma values, instead of just 1.0 (0)
+#define FULL_GAMMA true
+// Use all pixel formats, instead of just the ones that work
+#define FULL_PIXEL_FORMATS false
+// Also set the copy filter values for prev and next rows
+#define CHECK_PREV_AND_NEXT true
+
+struct CopyFilterTestContext
+{
+  PixelFormat pixel_fmt;
+  GammaCorrection gamma;
+  u8 prev_copy_filter_sum;
+  u8 copy_filter_sum;
+  u8 next_copy_filter_sum;
+  bool intensity_fmt;
+};
+template <>
+struct fmt::formatter<CopyFilterTestContext>
+{
+  constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
+  template <typename FormatContext>
+  auto format(const CopyFilterTestContext& test, FormatContext& ctx) const
+  {
+    return fmt::format_to(ctx.out(),
+                          "pixel_fmt: {}, gamma: {}, copy filter: {}/{}/{}, intensity: {}",
+                          test.pixel_fmt, test.gamma, test.prev_copy_filter_sum,
+                          test.copy_filter_sum, test.next_copy_filter_sum, test.intensity_fmt);
+  }
+};
+
+void SetCopyFilter(u8 prev_copy_filter_sum, u8 copy_filter_sum, u8 next_copy_filter_sum);
+
+GXTest::Vec4<u8> GenerateEFBColor(u16 x, u16 y)
+{
+  const u8 r = static_cast<u8>(x);
+  const u8 g = static_cast<u8>(y == 4 ? x : (y == 3 ? 255 : 0));
+  const u8 b = static_cast<u8>(x);
+  const u8 a = static_cast<u8>(x);
+  return {r, g, b, a};
+}
+
+static u32 GenerateEFBDepth(u16 x, u16 y)
+{
+  auto color = GenerateEFBColor(x, y);
+  return (u32(color.r) << 16) | (u32(color.g) << 8) | color.b;
+}
+
+static void SetPixelFormat(PixelFormat pixel_fmt)
+{
+  PEControl ctrl{.hex = BPMEM_ZCOMPARE << 24};
+  ctrl.pixel_format = pixel_fmt;
+  ctrl.zformat = DepthFormat::ZLINEAR;
+  ctrl.early_ztest = false;
+  CGX_LOAD_BP_REG(ctrl.hex);
+}
+
+static void FillEFB(PixelFormat pixel_fmt)
+{
+  // Don't set the format to Z24 here since we need RGB8 for our EFB copy+z-texture hack below
+  SetPixelFormat(pixel_fmt == PixelFormat::Z24 ? PixelFormat::RGB8_Z24 : pixel_fmt);
+
+  ZMode zmode{.hex = BPMEM_ZMODE << 24};
+  zmode.testenable = true;
+  zmode.func = CompareMode::Always;
+  zmode.updateenable = true;
+  CGX_LOAD_BP_REG(zmode.hex);
+
+  CGX_WaitForGpuToFinish();
+
+  CGX_PEPokeDither(false);
+  CGX_PEPokeAlphaUpdate(true);
+  CGX_PEPokeColorUpdate(true);
+  CGX_PEPokeBlendMode(GX_BM_NONE, SrcBlendFactor::Zero, DstBlendFactor::Zero, LogicOp::Set);
+  CGX_PEPokeAlphaRead(GX_READ_NONE);
+  CGX_PEPokeZMode(true, CompareMode::Always, true);
+
+  // For some reason GX_PokeARGB hangs when using this format
```

Comment:
It's possibly a hardware bug, or perhaps the hardware expects use of a 16-bit write and libogc doesn't do that.

I tried to fill the EFB in RGB8 mode and then switch to RGB565, but that didn't work.

<details><summary>Patch</summary>

```patch
diff --git a/gxtest/copyfilter.cpp b/gxtest/copyfilter.cpp
index 32dd050..b10e8ab 100644
--- a/gxtest/copyfilter.cpp
+++ b/gxtest/copyfilter.cpp
@@ -14,13 +14,13 @@
 #include "gxtest/util.h"
 
 // Use all copy filter values (0-63*3), instead of only 64
-#define FULL_COPY_FILTER_COEFS true
+#define FULL_COPY_FILTER_COEFS false
 // Use all gamma values, instead of just 1.0 (0)
-#define FULL_GAMMA true
+#define FULL_GAMMA false
 // Use all pixel formats, instead of just the ones that work
 #define FULL_PIXEL_FORMATS false
 // Also set the copy filter values for prev and next rows
-#define CHECK_PREV_AND_NEXT true
+#define CHECK_PREV_AND_NEXT false
 
 struct CopyFilterTestContext
 {
@@ -74,7 +74,10 @@ static void SetPixelFormat(PixelFormat pixel_fmt)
 static void FillEFB(PixelFormat pixel_fmt)
 {
   // Don't set the format to Z24 here since we need RGB8 for our EFB copy+z-texture hack below
-  SetPixelFormat(pixel_fmt == PixelFormat::Z24 ? PixelFormat::RGB8_Z24 : pixel_fmt);
+  if (pixel_fmt == PixelFormat::Z24 || pixel_fmt == PixelFormat::RGB565_Z16)
+    SetPixelFormat(PixelFormat::RGB8_Z24);
+  else
+    SetPixelFormat(pixel_fmt);
 
   ZMode zmode{.hex = BPMEM_ZMODE << 24};
   zmode.testenable = true;
@@ -91,10 +94,6 @@ static void FillEFB(PixelFormat pixel_fmt)
   CGX_PEPokeAlphaRead(GX_READ_NONE);
   CGX_PEPokeZMode(true, CompareMode::Always, true);
 
-  // For some reason GX_PokeARGB hangs when using this format
-  if (pixel_fmt == PixelFormat::RGB565_Z16)
-    return;
-
   for (u16 x = 0; x < 256; x++)
   {
     for (u16 y = 0; y < 8; y++)
@@ -223,6 +222,10 @@ static void FillEFB(PixelFormat pixel_fmt)
 
     SetPixelFormat(pixel_fmt);
   }
+  else if (pixel_fmt == PixelFormat::RGB565_Z16)
+  {
+    SetPixelFormat(pixel_fmt);
+  }
 }
 
 #if FULL_GAMMA
@@ -235,7 +238,7 @@ static const std::array<GammaCorrection, 1> GAMMA_VALUES = { GammaCorrection::Ga
 static const std::array<PixelFormat, 8> PIXEL_FORMATS = { PixelFormat::RGB8_Z24, PixelFormat::RGBA6_Z24, PixelFormat::RGB565_Z16, PixelFormat::Z24, PixelFormat::Y8, PixelFormat::U8, PixelFormat::V8, PixelFormat::YUV420 };
 #else
 // These formats work on Dolphin and on real hardware
-static const std::array<PixelFormat, 3> PIXEL_FORMATS = { PixelFormat::RGB8_Z24, PixelFormat::RGBA6_Z24, PixelFormat::Z24 };
+static const std::array<PixelFormat, 1> PIXEL_FORMATS = { PixelFormat::RGB565_Z16 };
 #endif
 
 // Applies to current row
@@ -462,10 +465,6 @@ void CopyFilterTest(const CopyFilterTestContext& ctx)
 
 void CheckEFB(PixelFormat pixel_fmt)
 {
-  // For some reason GX_PokeARGB hangs when using this format
-  if (pixel_fmt == PixelFormat::RGB565_Z16)
-    return;
-
   START_TEST();
 
   if (pixel_fmt != PixelFormat::Z24)
```

</details>

For the EFB peeks, the result was as if the EFB was in RGB8_Z24 mode (calling `CheckEFB(PixelFormat::RGB8_Z24)` results in a pass).  For the EFB copies, the result was 255 in all channels.

In any case, RGB565 is used for antialiased framebuffers with multiple samples, so it probably should be handled in a more complex manner than just setting all samples to the same value.  But dolphin doesn't implement this currently, so I'm not going to worry about it for now.

Reply:
> or perhaps the hardware expects use of a 16-bit write and libogc doesn't do that

Or a 64bit write with three 16bit samples?

> For the EFB peeks, the result was as if the EFB was in RGB8_Z24 mode (calling CheckEFB(PixelFormat::RGB8_Z24)

Yeah, that's what I've always suspected, that it did nothing special for color peaks in RGB565 mode and you would just get the raw bits as expected in 24 bit mode. 

> For the EFB copies, the result was 255 in all channels.

Woah... Something weird is going on here. We have always assumed you could just reinterpret the framebuffer bits without issue. Are there extra bits in the EFB for coverage that we don't know about?

------------

I'm fine for pushing this all to the future. Dolphin doesn't even implement RGB565 mode, except for converting the depth to a 16bit value on EFB peaks.